PVC Protection Alpha Feature E2E Tests

pospispa opened the pull request:
**What this PR does / why we need it**:
PVC Protection alpha feature was introduced in PRs:
- https://github.com/kubernetes/kubernetes/pull/55824
- https://github.com/kubernetes/kubernetes/pull/55873

This PR adds E2E tests for this feature.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
N/A

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

Commit:
PVC Protection Alpha Feature E2E Tests

PVC Protection alpha feature was introduced in PRs:
- https://github.com/kubernetes/kubernetes/pull/55824
- https://github.com/kubernetes/kubernetes/pull/55873

That's why E2E tests for this feature are added.

pospispa commented:
@msau42 @jsafrane PTAL

/kind technical-debt
/sig storage
/assign @msau42 
/assign @jsafrane

pospispa commented:
/retest

gnufied commented on test/e2e/storage/pvc_protection.go:
```diff
@@ -0,0 +1,134 @@
+		suffix := "pvc-protection"
+		storageClass = newStorageClass(testStorageClass, nameSpace, suffix)
+		storageClass, err = client.StorageV1().StorageClasses().Create(storageClass)
+		Expect(err).NotTo(HaveOccurred(), "Error creating storage class")
```
Any particular reason this test creates new storage class rather than using the one that already exists?

pospispa replied:
I don't know that one already exist.
Is it a good idea to assume that a Storage Class exist?

gnufied replied:
There should already be a default storageclass available in all environments. I think plenty of tests assume that, so we would be fine.

gnufied commented on test/e2e/storage/pvc_protection.go:
```diff
@@ -0,0 +1,134 @@
+	})
+
+	AfterEach(func() {
+		framework.ExpectNoError(client.StorageV1().StorageClasses().Delete(storageClass.Name, metav1.NewDeleteOptions(0)))
```
The test should delete the pvc in question here because multiple `It` blocks are using same PVC object. I know that, both `It` blocks happen to delete the PVC in their tests but a premature exit condition or something else may cause PVC to be not deleted between different `It` blocks.

pospispa replied:
So if I understand correctly, you're saying that in the `AfterEach` must be something like:
```
if pvc != nil {
    delete(pvc)
}
```
or can I rely on the `framework` that will delete the namespace anyway?

pospispa commented:
/sig testing

gnufied commented on test/e2e/storage/pvc_protection.go:
```diff
@@ -0,0 +1,134 @@
+	return fmt.Errorf("PersistentVolumeClaim %s is not removed from the system within %v", pvcName, timeout)
+}
+
+func skipUnlessPVCProtectionEnabled() {
```
Does this work externally in a already running cluster? What I understand is - once a cluster is already running with feature flag enabled, the e2e test obviously runs externally and the e2e process does not know, if the feature flag is enabled or not. May be there is some magic that makes it work with e2e job but it does not work locally for me. 

I also noticed that some of the tests that @msau42  introduced uses similar feature gating in e2e tests.

pospispa replied:
I "copied" the feature gating from @msau42 

The feature gating doesn't work externally when I try it on my notebook. I have no idea if it works in e2e jobs.

msau42 replied:
I don't recall having feature gating in my e2e tests.  This won't work as there's no way to externally check if a feature gate is enabled in a cluster.

The way to add e2e tests for alpha features is to:
* Add a `[Feature: blah]` tag to your suite
* In the test infra config, update the alpha job to include your tag added above.

pospispa replied:
@msau42 @gnufied thank you for advice. I removed the feature gating.

pospispa commented:
/retest

pospispa commented:
@gnufied @msau42 I believe I've addressed all your comments. Please, let me know whether the PR looks good to you and whether I should squash the commits.

msau42 commented on test/e2e/storage/pvc_protection.go:
```diff
@@ -0,0 +1,125 @@
+		err = client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(pvc.Name, metav1.NewDeleteOptions(0))
+		Expect(err).NotTo(HaveOccurred(), "Error deleting PVC")
+
+		By("Checking that the PVC status is Terminating")
```
Would it be worth it to wait a little bit to make sure it stays Terminating?

pospispa replied:
@msau42 it's a good question. Actually, I plan to extend this test after [changes into scheduler](https://github.com/kubernetes/kubernetes/pull/55957) are merged. After the extension the test will check that the PVC is Terminating, then it will try to start a pod that uses the Terminating PVC, check that the pod is not scheduled, delete the not-scheduled pod and then verify that the PVC status is still Terminating. So it will make sure that the PVC stays Terminating for some time.

msau42 commented:
Please go ahead and squash

msau42 reviewed (commented):
/lgtm

msau42 commented:
/approve no-issue

jsafrane commented on test/e2e/storage/pvc_protection.go:
```diff
@@ -0,0 +1,125 @@
+})
+
+// waitForPersistentVolumeClaimBeRemoved waits for a PersistentVolumeClaim to be removed from the system until timeout occurs, whichever comes first.
+func waitForPersistentVolumeClaimBeRemoved(c clientset.Interface, ns string, pvcName string, Poll, timeout time.Duration) error {
```
It would be nice to move this to `test/e2e/framework/util.go`, maybe in the aforementioned PR with scheduler tests.

pospispa replied:
Should private func that is used only on one place be made public?

jsafrane replied:
It won't hurt. I like them all together so I can see what `WaitFor<something>` are available when I write my test. It's harder to find when it's just in `pvc_protection.go`.

pospispa commented on test/e2e/storage/pvc_protection.go:
```diff
@@ -0,0 +1,125 @@
+		err = framework.DeletePodWithWait(f, client, pod)
+		Expect(err).NotTo(HaveOccurred(), "Error terminating and deleting pod")
+
+		By("Checking that the PVC is automativally removed from the system because it's no longer in active use by a pod")
```
fixed typo.
@jsafrane please, will you LGTM?

pospispa commented:
/retest

jsafrane commented:
/lgtm

pospispa commented:
/test pull-kubernetes-unit

pospispa commented:
@msau42 @jsafrane @gnufied please, will you LGTM the PR after rebase. Thank you.

gnufied commented:
/lgtm
